File: apps/api-rs/src/document/chunking/table.rs
use super::*;
pub(super) fn split_table_group(
    file_type: FileType,
    kb_id: Uuid,
    parse_job_id: Uuid,
    group: &BlockGroup,
    cfg: &ChunkConfig,
    chunks: &mut Vec<ChunkDraft>,
) -> bool {
    let Some(block) = group.blocks.first() else {
        return false;
    };
    let lines = block.cleaned_text.lines().collect::<Vec<_>>();
    if lines.len() <= 2
        || (lines.len().saturating_sub(2) <= cfg.max_table_rows_per_chunk
            && group.tokens <= cfg.max_table_token_per_chunk)
    {
        return false;
    }

    let header = &lines[..2];
    let body = &lines[2..];
    let mut current_rows = Vec::new();
    for row in body {
        let mut candidate = header.to_vec();
        candidate.extend(current_rows.iter().copied());
        candidate.push(row);
        let exceeds_rows = current_rows.len() >= cfg.max_table_rows_per_chunk;
        let exceeds_tokens = estimate_tokens(&candidate.join("\n")) > cfg.max_table_token_per_chunk;
        if !current_rows.is_empty() && (exceeds_rows || exceeds_tokens) {
            push_table_chunk(
                file_type,
                kb_id,
                parse_job_id,
                block,
                header,
                &current_rows,
                chunks,
            );
            current_rows.clear();
        }
        current_rows.push(*row);
    }
    if !current_rows.is_empty() {
        push_table_chunk(
            file_type,
            kb_id,
            parse_job_id,
            block,
            header,
            &current_rows,
            chunks,
        );
    }
    true
}

fn push_table_chunk(
    file_type: FileType,
    kb_id: Uuid,
    parse_job_id: Uuid,
    block: &CleanedBlock,
    header: &[&str],
    rows: &[&str],
    chunks: &mut Vec<ChunkDraft>,
) {
    let mut part = block.clone();
    part.cleaned_text = header
        .iter()
        .chain(rows.iter())
        .copied()
        .collect::<Vec<_>>()
        .join("\n");
    chunks.push(chunk_from_group(
        file_type,
        kb_id,
        parse_job_id,
        single_block_group(part),
        "table_rows",
    ));
}

File: apps/api-bun/src/document/chunking/table.ts
// 移植自 apps/api-rs/src/document/chunking/table.rs —— 行为对齐 Rust 原版

// 单块表格切分：按行拆成多片，每片重复表头，并写入 table_part_index

import type { ChunkConfig } from '../chunking.ts';
import { chunkFromGroup, singleBlockGroup, type BlockGroup } from '../chunking.ts';
import type { ChunkDraft, FileType } from '../types.ts';
import { splitTableText } from './postprocess.ts';

export function splitTableGroup(
  fileType: FileType,
  kbId: string,
  parseJobId: string,
  group: BlockGroup,
  cfg: ChunkConfig,
  chunks: ChunkDraft[],
): boolean {
  const block = group.blocks[0];
  if (block === undefined) return false;
  const parts = splitTableText(block.cleaned_text, cfg);
  if (parts.length <= 1) return false;
  for (let partIndex = 0; partIndex < parts.length; partIndex += 1) {
    const part = { ...block, cleaned_text: parts[partIndex]! };
    const chunk = chunkFromGroup(fileType, kbId, parseJobId, singleBlockGroup(part), 'table_rows');
    if (isRecord(chunk.metadata)) {
      chunk.metadata['table_part_index'] = partIndex;
    }
    chunks.push(chunk);
  }
  return true;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
